feat(gossipsub): introduce backpressure

jxs opened the pull request:
Follows @mxinden implementation suggestion on https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1811241573

## Notes & open questions
Publish, GRAFT, PRUNE and Subscription messages are prioritized, Publish messages are cap'ed, using `sender.clone().try_send` doesn't work as the queue is shared by the `Receiver`. I used an `AtomicUsize` for that.
Used `async-channel` to be able to check if `is_empty()` in the `ConnectionHandler`.
Reused the `InsufficientPeers` peers `Error` as to not introduce another variant and make the changes breaking, but this is probably a breaking change in the sense that the internal behaviour is changing no?
if this design makes sense I can then add a test to test queues fill.
<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

jxs committed:
feat(gossipsub): introduce backpressure

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -332,6 +335,9 @@ pub struct Behaviour<D = IdentityTransform, F = AllowAllSubscriptionFilter> {
     metrics: Option<Metrics>,
+
+    /// Connection handler message queue channels.
+    handler_send_queues: HashMap<PeerId, RpcSender>,
```
I think we might have to make this by `ConnectionId`. Or make the value of the hashmap a `Vec` of `RpcSender`.

jxs replied:
I see, because `ConnectionHandler`s may send to different `PeerId`s during its existence right? But then how do we relate the mesh peers which are identified by `PeerId`s to `ConnectionId`s?

mxinden replied:
> because `ConnectionHandler`s may send to different `PeerId`s during its existence right?

No. A `ConnectionHandler` only ever connects the local node to a single remote node (i.e. remote `PeerId`) during its existence. Though note that the local node might have multiple `ConnectionHandler`s to a single remote node, each with a different `ConnectionId`.

I think the problem that Thomas is hinting at, is that we might have multiple connections to a single `PeerId`. In other words, there might be multiple `RpcSender`s per `PeerId`.

jxs replied:
Thanks for explaining Max, If I understand correctly it's also why Thomas mentioned that we'd need to distinguish between `Full` and `Closed` when sending the message to the `Receiver` so that with the idea of a `Vec` of `RpcSender` we try all the `ConnectionHandlers` right? 
If that is so, since `async-channel` is mpmc ( where each message can be received by only one of all existing consumers) why don't we clone the `RpcReceiver` so that the first `ConnectionHandler` reading the message sends it and we assure the channel never gets closed? Presented this idea on dd13fcd6f8a0980996c2ea9c947df402726954c5

mxinden replied:
With the latest refactorings, is this field still needed? The `RpcSender` is now part of `PeerConnections`, right?

jxs replied:
no no, thanks Max updated!

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -512,3 +517,99 @@ impl fmt::Display for PeerKind {
+        // The priority queue is then polled.
+        if let Poll::Ready(rpc) = Pin::new(&mut self.priority).poll_next(cx) {
+            return Poll::Ready(rpc);
+        }
```
Why we are polling `self.priority` twice?

jxs replied:
thanks Thomas updated, leftovers from experimenting with a third queue for `ControlAction` messages.

thomaseizinger commented on protocols/gossipsub/src/config.rs:
```diff
@@ -782,6 +789,10 @@ impl ConfigBuilder {
         self
     }
 
+    pub fn connection_handler_queue_len(&mut self, len: usize) {
+        self.config.connection_handler_queue_len = len;
+    }
```
Why does this need to be configurable?

jxs replied:
We do not know yet the optimal value for this, and it should be different to different usecases and hardware specs. A bigger limit allows for more memory to be allocated and therefore more throughput, OTOH a smaller limit allows for less memory usage at the cost of a reduced throughput.

mxinden replied:
> it should be different to different usecases and hardware specs

To some extend. Though I would assume that we can cover most use-cases with a reasonable default. Choosing this value too small results in task switches between the `NetworkBehaviour` (i.e. the main `Swarm::poll` task) and the `ConnectionHandler` task (i.e. the connection task), given that the former is blocked on the latter making progress. I would expect a task switch to be smaller than a micro-second. Contrasting this to network latencies that operate on the order of milliseconds. This is at least a factor of 1_000. Thus I doubt optimizing the queue length beyond a reasonable default will yield any noticeable return.

Long story short, experimentation sounds good to me, though once we found a value, I doubt we need the configuration flag.

AgeManning replied:
We have done some experimentation. 

Lighthouse operates in bursts. When we initially start, we need to subscribe to all the topics (we have a few, order few hundred). We would GRAFT/PRUNE to get into a stable mesh. We would then receive lots of IHAVE's for past messages we have not seen and attempt to download those with IWANT. 

Some values we saw were around the 4k mark. 

To handle bursts etc, I was in favour of a time-bound. Especially because gossipsub is very time-dependent. We shouldn't care about sending a message that was supposed to be sent a minute ago. 

I've spoken with @jxs and the middle ground seems to be we have a space-bound channel for important messages so the behaviour can know if a peer is being over-burdened.  

Then we time-bound the forward/IHAVE/IWANT messages which should hopefully help free-up the queue.

If we go this approach without a configurable parameter, I'd opt for a high value which only gets hit in the case that a peer piles up messages and go with something of the order 5k (for users who want to send lots of messages in bursts).

mxinden replied:
> If we go this approach without a configurable parameter, I'd opt for a high value which only gets hit in the case that a peer piles up messages and go with something of the order 5k (for users who want to send lots of messages in bursts).

Handling bursts with a queue (in our case, the queue is in the channel) sounds good to me. Using a high value, i.e. a value that is not hit in the happy path, is also fine by me.

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -512,3 +517,99 @@ impl fmt::Display for PeerKind {
+/// Create `RpcOut` channel that is priority aware.
+pub(crate) fn rpc_channel(cap: usize) -> (RpcSender, RpcReceiver) {
+    let (priority_sender, priority_receiver) = async_channel::unbounded();
+    let (non_priority_sender, non_priority_receiver) = async_channel::bounded(cap / 2);
```
`async-channel` is an MPMC channel but we don't ever `.clone()` the `Receiver`, right? So why not use the `mpsc` channel from `futures`?

thomaseizinger replied:
The `futures::channel::mpsc` implementation gives you one slot per `Sender` + whatever capacity you initialize the channel with. So by doing `sender.clone().try_send`, I _think_ we are guaranteed to be able to send a message, even if the `Sender` is immediately dropped after. Technically, that makes the channel unbounded but if we only do it for `Control` messages, that shouldn't matter much because they are so small.

I think that is what @mxinden meant in https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1811241573.

jxs replied:
>async-channel is an MPMC channel but we don't ever .clone() the Receiver, right? So why not use the mpsc channel from futures?

because we need to check if the channel is empty on the connection handler when determining if we need to create the outbound stream:
```rust
        if !self.send_queue.is_empty()
            && self.outbound_substream.is_none()
            && !self.outbound_substream_establishing
        {
```
`handler.rs:231`

thomaseizinger replied:
You can wrap the `Receiver` in a `Peekable` and check if it has an item: https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.peekable

jxs replied:
thanks Thomas I wasn't aware of this! Nor of the property Max mentioned regarding the `Sender` clone. See my question regarding MPMC on https://github.com/libp2p/rust-libp2p/pull/4914#discussion_r1405212999

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -537,7 +544,8 @@ where
             let event = RpcOut::Subscribe(topic_hash.clone());
-            self.send_message(peer, event);
+            self.send_message(peer, event)
+                .expect("Subscribe messages should be always sent");
```
I think it would be useful if we'd split `RpcOut` to avoid this. Something like:

- `ControlRpcOut`
- `Publish`
- `Forward`

And then we can have:

```rust
fn send_control_message(); 					// Never fails
fn send_publish() -> Result<(), ...>;		// Can fail
fn send_forward(); 							// Never fails but pushes into a different channel
```

jxs replied:
yeah thought about it, but `ControlAction` has different types of priorities, IHAVE/IWANT are low and GRAFT/PRUNE are high priority, all of them are under the `ControlAction` enum as struct variants. We could split them into isolated structures to then use both on `PriorityRpcOut` and and `RpcOut` so that we could do:
```rust
fn prune(&mut self, graft: Graft);
fn graft(&mut self, graft: Graft);
```

but `ControlAction` is pub so that would be a breaking change

mxinden replied:
> but `ControlAction` has different types of priorities, IHAVE/IWANT are low and GRAFT/PRUNE are high priority

Why is this distinction relevant?

The goal of this pull request is to prevent unbounded growth of `send_queue`. Neither IHAVE/IWANT nor GRAFT/PRUNE are large. Correct me if I am wrong. Thus differentiating among the two is not relevant for this pull request.

jxs replied:
I was following @AgeManning on https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1793933201 but you are probably right Max, updated the code to implement Thomas suggested methods into `RpcSender` which allow us to not split `RpcOut` and meanwhile also remove `send_message` and the double iterations on `Behaviour`, ptal Thomas

AgeManning replied:
I think the distinction is important. 

GRAFT/PRUNE are very important in maintaining the overlay mesh network. If we drop these messages we could break the whole network, i.e nodes could start randomly having broken links and will not know to find new mesh peers and messages can stop being forwarded through the network. 

IHAVE/IWANT are very low priority. In fact if we are struggling with send/receiving messages we probably want to drop these along with forward messages. Although the messages themselves are fairly small (but we can send a large number of message-ids, like a few thousand (configurable)), they both induce a greater message load.  IHAVE can induce IWANT requests from other peers, which ends up making us publish more messages (which we don't want if we are struggling).   IWANT  will request more messages to be sent to us and further consume bandwidth. An IWANT can request a fair amount more messages (up to 500 in lighthhouse). 

For these reasons, I think if we are in the unhappy case where we are struggling, we want to drop IHAVE/IWANT and forward messages, but we definitely do not want to drop GRAFT/PRUNE if we can help it.

AgeManning replied:
As I write this, we probably also want to make sure that messages that we respond to an IWANT with, gets sent as a "forward" message and not a "publish" message. 
If it is sent too late, we will have already broken our promise, as they are time-bound anyway. We want to remove these from sending if we can't send them in time.

mxinden replied:
Distinguishing in GRAFT/PRUNE and IHAVE/IWANT makes sense to me. Thank you for expanding on it.

That said, I suggest not enlarging the scope of this pull request. I suggest to fix the unbounded `send_queue` problem in this pull request (https://github.com/libp2p/rust-libp2p/issues/4667). Once merged I suggest following up with one or more pull requests introducing various optimizations, e.g. time based prioritization of forward messages, differentiation of GRAFT/PRUNE and IHAVE/IWANT, ...

I won't block this pull request in case you want to do all of this in one. Though I expect separate pull requests to be significantly either to implement, review and debug.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2730,19 +2752,24 @@ where
+        if sender.try_send(rpc.clone()).is_err() {
+            tracing::debug!(peer=%peer_id, "Dropping message as peer is full");
+            return Err(());
+        }
```
We should differentiate between `Closed` and `Full` here and perhaps return the message back to the caller in `Err`.

jxs replied:
but `Closed` is unreachable no?

thomaseizinger replied:
Not necessarily. A connection runs in a different task and on a multi-threaded executor, this connection could get dropped and thus the receiver freed before we receive the event that the connection is closed (where we'd clean up the state).

It might not matter though for this case so probably fine :)

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2730,19 +2752,24 @@ where
         if let Some(m) = self.metrics.as_mut() {
             if let RpcOut::Publish(ref message) | RpcOut::Forward(ref message) = rpc {
                 // register bytes sent on the internal metrics.
                 m.msg_sent(&message.topic, message.raw_protobuf_len());
```
If you compute this before we use `try_send`, you can avoid the `.clone()` on the message.

jxs replied:
but if the send fails when we are adding wrong metrics data as the message was not actually sent no?

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2730,19 +2752,24 @@ where
+        let sender = self
+            .handler_send_queues
+            .get_mut(&peer_id)
+            .expect("Peerid should exist");
```
Are we sure this can never be hit? I'd rather have us return an error here.

jxs replied:
not 100% sure yeah, although we are removing it on disconnects, but then shouldn't we log it here instead? Also if we are going to use `ConnectionId`s  as you suggested on [above](https://github.com/libp2p/rust-libp2p/pull/4914/files#r1402796187) the question I asked there is also valid here, how and where do we related `PeerId`'s with `ConnectionId`s

thomaseizinger commented on protocols/gossipsub/src/behaviour/tests.rs:
Ugh, this is so painful ...

We should really rewrite these tests to go against the `Swarm` API using `libp2p-swarm-test`. Changing tests AND implementation in one PR makes me a lot less confident in a green CI.

There are a lot of tests to change. Can we perhaps make this a group effort? (cc @mxinden @AgeManning)

It probably makes sense to first rewrite a few in order to create some utility functions for setting up networks. But once those are in place, perhaps we can chip away at them together? There are 85 tests to change ..

thomaseizinger replied:
I also don't want to blow the scope of this work too much. It is just that reviewing all the changes to these tests also takes a lot of time but it is an effort that we aren't going to benefit from in the future.

jxs replied:
Yeah agree Thomas, let me look into that

AgeManning replied:
Yeah. I'm happy to help out also. My opinion is to try and get these changes in first (we'll be doing fairly extensive tests on live networks as well as small scale simulations) to give us confidence in the changes. 
Then in a future PR we re-write all the tests.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -250,10 +247,11 @@ impl EnabledHandler {
-                        self.send_queue.shrink_to_fit();
-                        self.outbound_substream =
-                            Some(OutboundSubstreamState::PendingSend(substream, message));
+                    if let Poll::Ready(Some(message)) = self.send_queue.poll_next_unpin(cx) {
```
Can we now get rid of `OutboundSubstreamState` and just use an `async move` with a `loop` inside? I guess that would require us to be able to `.clone()` the `Receiver` so we can keep one around in case the stream fails and we want to re-establish it.

Probably not worth doing as part of this to keep the scope small but it would be good to debate, why we are retrying to establish the stream if it failed. Any form of loop that retries something is always a bit sus in my eyes. We could as well rely on an upper layer to re-establish the connection. I don't see why trying 5 times to re-establish the stream is fundamentally different to just disabling ourselves after the first one.

jxs replied:
Yeah agree we should try it on a subsequent PR

thomaseizinger reviewed (commented):
Great start! Left some comments! :)

jxs committed:
remove duplicated poll from RpcReceiver

jxs committed:
make RpcReceiver part of RpcSender,

and clone it per ConnectionHandler, this will allow us to always have an open Receiver.

jxs committed:
split send_message into multiple RpcSender methods,

to allow for better handling of each message send.

jxs reviewed (commented):
did not forget about Thomas remark on https://github.com/libp2p/rust-libp2p/pull/4914/files#r1402807803. I am just trying to close the api so that we know what we need from the updated tests

jxs committed:
cargo clippy

jxs committed:
add CHANGELOG.md entry

jxs committed:
update to use futures::channel and connected_peers

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into gossipsub-backpressure-cont

mxinden commented on protocols/gossipsub/CHANGELOG.md:
```diff
@@ -1,7 +1,11 @@
-## 0.46.1
+## 0.46.1 - unreleased
+
+- Implement backpressure by diferentiating between priority and non priority messages.
```
```suggestion
- Implement backpressure by differentiating between priority and non priority messages.
```

mxinden commented on protocols/gossipsub/Cargo.toml:
```diff
@@ -37,12 +37,13 @@ sha2 = "0.10.8"
 smallvec = "1.11.2"
 tracing = "0.1.37"
 void = "1.0.2"
+async-channel = "1.9.0"
```
Still needed?

jxs replied:
no no, thanks Max, updated

mxinden commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2810,8 +2839,14 @@ where
+            .connected_peers
+            .get_mut(&peer_id)
+            .expect("Peerid should exist")
+            .clone();
```
Why is this clone needed?

jxs replied:
it's not, thanks Max!

mxinden commented on protocols/gossipsub/src/types.rs:
```diff
@@ -71,12 +77,115 @@ impl std::fmt::Debug for MessageId {
+    /// Send a `RpcOut::Subscribe` message to the `RpcReceiver`
+    /// this is high priority. By cloning `futures::channel::mpsc::Sender`
+    /// we get one extra slot in the channel's capacity.
+    pub(crate) fn subscribe(&mut self, topic: TopicHash) {
```
Should `subscribe` return a `Result` for the case where all channels to the remote peer are closed?

jxs replied:
Shouldn't we assume that situation as unreachable? i.e. there will always be at least one connection id per peer? as we add them and remove them on new and closed connections.

If not all `PeerConnections` send methods should account for that right?

mxinden replied:
See also https://github.com/libp2p/rust-libp2p/pull/4914#discussion_r1431207894.

mxinden commented on protocols/gossipsub/src/types.rs:
```diff
@@ -71,12 +77,115 @@ impl std::fmt::Debug for MessageId {
+                }
+            }
+        }
+        unreachable!("At least one peer should be available");
```
Why is this unreachable? Say that all channels to the given peer are closed, the `for` loop above will not `return` and thus this `unreachable` is hit, no?

jxs replied:
yeah, it's as I asked above, isn't the situation of having all channels to a given peer closed `unreachable`? 
On `handle_established_{inbound,outbound}_connection` we add the `PeerId` and the `ConnectionId` to `connected_peers`, which is then removed on `on_connection_closed` removing the `ConnectionId`, and the `PeerId` if  `remaining_established_ is zero.

mxinden replied:
Say that each connection has an error, e.g. here:

https://github.com/libp2p/rust-libp2p/blob/c3500bb79a57fde067394911bd2cde76c403da9b/swarm/src/connection/pool/task.rs#L237

Then each connection is going to close their channel.

Say that "the same" time (concurrently) the user calls `Behaviour::publish`.

In such case, the `Behaviour` would still track the connections, even though each channel would be dropped.

Does the above make sense?

mxinden commented on protocols/gossipsub/src/types.rs:
```diff
@@ -71,12 +77,115 @@ impl std::fmt::Debug for MessageId {
+                    }
+                    return;
+                }
+                Err(err) if err.is_full() => {
+                    tracing::trace!("Queue is full, dropped Forward message");
+                    return;
+                }
+                // Channel is closed, try another sender.
+                Err(err) => {
+                    rpc = err.into_inner();
+                }
```
Still have to give this more thought, though of the top of my head, trying a different connection in both cases is fine, no?

mxinden commented on protocols/gossipsub/src/types.rs:
```diff
@@ -512,3 +621,45 @@ impl fmt::Display for PeerKind {
+    pub(crate) non_priority: Sender<RpcOut>,
+}
+
+/// `RpcOut` sender that is priority aware.
```
```suggestion
/// `RpcOut` receiver that is priority aware.
```

jxs replied:
thanks Max, addressed!

jxs committed:
address review

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into gossipsub-backpressure-cont

jxs committed:
improve wording

jxs committed:
clippy

jxs committed:
update cargo.toml
